pyphasefield/pyphasefield/field.py

```diff
@@ -24,41 +28,49 @@ def __array_finalize__(self, obj):
         if obj is None: return
         self._name = getattr(obj, '_name', "phi")
         self._field_type = getattr(obj, '_field_type', "scalar")
-        self._cell_spacing = getattr(obj, '_cell_spacing', 0.0000046)
-        self._dimensions_of_simulation_region = getattr(obj, '_dimensions_of_simulation_region', [200,200])
-        self._number_of_dimensions = getattr(obj, '_number_of_dimensions', 2)
+        self._simulation = getattr(obj, '_simulation', defaultsim)
         
     def gradient_cell(self):
-        array = (np.roll(self, -1, 0) - np.roll(self, 1, 0))/(2*self._cell_spacing)
+        number_of_dimensions = len(self.shape)
+        inverse_cell_spacing = 1./(self._simulation._cell_spacing_in_cm)
+        inverse_2dx = 0.5*inverse_cell_spacing
+        array = (np.roll(self, -1, 0) - np.roll(self, 1, 0))*inverse_2dx
         array = array[np.newaxis] #add dimension for different partial derivatives
-        for i in range(1, self._number_of_dimensions):
-            temp = (np.roll(self, -1, i) - np.roll(self, 1, i))/(2*self._cell_spacing)
+        for i in range(1, number_of_dimensions):
+            temp = (np.roll(self, -1, i) - np.roll(self, 1, i))*inverse_2dx
             temp = temp[np.newaxis]
             array = np.concatenate((array, temp), axis=0)
         return array
         
     def gradient_face_left(self):
-        array = (self - np.roll(self, 1, 0))/(self._cell_spacing)
+        number_of_dimensions = len(self.shape)
+        inverse_cell_spacing = 1./(self._simulation._cell_spacing_in_cm)
+        array = (self - np.roll(self, 1, 0))*inverse_cell_spacing
         array = array[np.newaxis] #add dimension for different partial derivatives
-        for i in range(1, self._number_of_dimensions):
-            temp = (self - np.roll(self, 1, i))/(self._cell_spacing)
+        for i in range(1, number_of_dimensions):
+            temp = (self - np.roll(self, 1, i))*inverse_cell_spacing
             temp = temp[np.newaxis]
             array = np.concatenate((array, temp), axis=0)
         return array
     
     def gradient_face_right(self):
-        array = (np.roll(self, -1, 0) - self)/(self._cell_spacing)
+        number_of_dimensions = len(self.shape)
+        inverse_cell_spacing = 1./(self._simulation._cell_spacing_in_cm)
+        array = (np.roll(self, -1, 0) - self)*inverse_cell_spacing
         array = array[np.newaxis] #add dimension for different partial derivatives
-        for i in range(1, self._number_of_dimensions):
-            temp = (np.roll(self, -1, i) - self)/(self._cell_spacing)
+        for i in range(1, number_of_dimensions):
+            temp = (np.roll(self, -1, i) - self)*inverse_cell_spacing
             temp = temp[np.newaxis]
             array = np.concatenate((array, temp), axis=0)
         return array
```

Comment:
Is this something that the user has to choose when developing an engine?

Reply:
Yes, both face_gradient functions essentially do the same thing, but are mirrored. Both return an array one dimension larger than the original, with each value along this dimension representing which piece of the gradient vector it corresponds to. face_gradient_right computes the face gradient using the central cell and the positive neighbor along each dimension, face_gradient_left is the central cell and the negative neighbor along each dimension